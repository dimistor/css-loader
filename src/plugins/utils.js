import postcss from 'postcss';

const createICSSImports = (imports) =>
  Object.keys(imports).map((path) => {
    const aliases = imports[path];
    const declarations = Object.keys(aliases).map((key) =>
      postcss.decl({
        prop: key,
        value: aliases[key],
        raws: { before: '\n  ' },
      })
    );

    return postcss
      .rule({
        selector: `:import(${path})`,
        raws: { after: '\n' },
      })
      .append(declarations);
  });

export default { createICSSImports };
