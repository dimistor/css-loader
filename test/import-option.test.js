import webpack from './helpers/compiler';
import evaluated from './helpers/evaluated';

import { normalizeErrors } from './helpers/utils';

describe('import', () => {
  test('true', async () => {
    const config = {
      loader: {
        options: {
          import: true,
        },
      },
    };
    const stats = await webpack('import/import.css', config);
    const { modules } = stats.toJson();
    const [, , , , , , , module] = modules;

    expect(module.source).toMatchSnapshot('module');
    expect(evaluated(module.source, modules)).toMatchSnapshot(
      'module (evaluated)'
    );
    expect(normalizeErrors(stats.compilation.warnings)).toMatchSnapshot(
      'warnings'
    );
    expect(normalizeErrors(stats.compilation.errors)).toMatchSnapshot('errors');
  });

  test('false', async () => {
    const config = {
      loader: {
        options: {
          import: false,
        },
      },
    };
    const stats = await webpack('import/import.css', config);
    const { modules } = stats.toJson();
    const [, module] = modules;

    expect(module.source).toMatchSnapshot('module');
    expect(evaluated(module.source, modules)).toMatchSnapshot(
      'module (evaluated)'
    );
    expect(stats.compilation.warnings).toMatchSnapshot('warnings');
    expect(stats.compilation.errors).toMatchSnapshot('errors');
  });
});
