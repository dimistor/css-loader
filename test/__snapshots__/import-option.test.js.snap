// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`import false: errors 1`] = `Array []`;

exports[`import false: module (evaluated) 1`] = `
Array [
  Array [
    1,
    "
@import url(test.css);
@import url('test.css');
@import url(\\"test.css\\");
@IMPORT url(test.css);
@import URL(test.css);
@import url();
@import url('');
@import url(\\"\\");
@import \\"test.css\\";
@import '';
@import \\"\\";
@import \\"   \\";
@import url();
@import url('');
@import url(\\"\\");
@import url(test.css) screen and print;
@import url(test.css)screen and print;
@import url(test-media.css) screen and print;
@import url(test-other.css) (min-width: 100px);
@import url(http://example.com/style.css);
@import url(http://example.com/other-style.css) screen and print;
@import url(\\"//example.com/style.css\\");
@import url(~package/test.css);
@import ;
@import foo-bar;
@import-normalize;
@import url('http://') :root {}
@import url('query.css?foo=1&bar=1');
@import url('other-query.css?foo=1&bar=1#hash');

.class {
  a: b c d;
}

.foo {
  @import 'path.css';
}
",
    "",
  ],
]
`;

exports[`import false: module 1`] = `
"// CSS runtime
module.exports = exports = require(\\"../../../src/runtime/api.js\\")(false);

// CSS module
exports.push([module.id, \\"\\\\n@import url(test.css);\\\\n@import url('test.css');\\\\n@import url(\\\\\\"test.css\\\\\\");\\\\n@IMPORT url(test.css);\\\\n@import URL(test.css);\\\\n@import url();\\\\n@import url('');\\\\n@import url(\\\\\\"\\\\\\");\\\\n@import \\\\\\"test.css\\\\\\";\\\\n@import '';\\\\n@import \\\\\\"\\\\\\";\\\\n@import \\\\\\"   \\\\\\";\\\\n@import url();\\\\n@import url('');\\\\n@import url(\\\\\\"\\\\\\");\\\\n@import url(test.css) screen and print;\\\\n@import url(test.css)screen and print;\\\\n@import url(test-media.css) screen and print;\\\\n@import url(test-other.css) (min-width: 100px);\\\\n@import url(http://example.com/style.css);\\\\n@import url(http://example.com/other-style.css) screen and print;\\\\n@import url(\\\\\\"//example.com/style.css\\\\\\");\\\\n@import url(~package/test.css);\\\\n@import ;\\\\n@import foo-bar;\\\\n@import-normalize;\\\\n@import url('http://') :root {}\\\\n@import url('query.css?foo=1&bar=1');\\\\n@import url('other-query.css?foo=1&bar=1#hash');\\\\n\\\\n.class {\\\\n  a: b c d;\\\\n}\\\\n\\\\n.foo {\\\\n  @import 'path.css';\\\\n}\\\\n\\", \\"\\"]);
"
`;

exports[`import false: warnings 1`] = `Array []`;

exports[`import true: errors 1`] = `Array []`;

exports[`import true: module (evaluated) 1`] = `
Array [
  Array [
    2,
    ".test{
  a: a;
}
",
    "screen and print",
  ],
  Array [
    4,
    ".test{
  a: b;
}
",
    "((min-width: 100px)) and (screen and print)",
  ],
  Array [
    3,
    ".test {
  c: d;
}
",
    "screen and print",
  ],
  Array [
    1,
    "@import url(http://example.com/style.css);",
    "",
  ],
  Array [
    1,
    "@import url(http://example.com/other-style.css);",
    "screen and print",
  ],
  Array [
    1,
    "@import url(//example.com/style.css);",
    "",
  ],
  Array [
    5,
    ".test{
  a: b
}
",
    "",
  ],
  Array [
    6,
    ".query {
  color: green;
}
",
    "",
  ],
  Array [
    7,
    ".other-query {
  color: green;
}
",
    "",
  ],
  Array [
    1,
    "@import url();
@import url('');
@import url(\\"\\");
@import '';
@import \\"\\";
@import \\"   \\";
@import url();
@import url('');
@import url(\\"\\");
@import ;
@import foo-bar;
@import-normalize;
@import url('http://') :root {}

.class {
  a: b c d;
}

.foo {
  @import 'path.css';
}
",
    "",
  ],
]
`;

exports[`import true: module 1`] = `
"// CSS runtime
module.exports = exports = require(\\"../../../src/runtime/api.js\\")(false);

// CSS imports
exports.i(require(\\"-!../../../src/index.js??ref--4!./test.css\\"), \\"screen and print\\");
exports.i(require(\\"-!../../../src/index.js??ref--4!./test-media.css\\"), \\"screen and print\\");
exports.i(require(\\"-!../../../src/index.js??ref--4!./test-other.css\\"), \\"(min-width: 100px)\\");
exports.push([module.id, \\"@import url(http://example.com/style.css);\\", \\"\\"]);
exports.push([module.id, \\"@import url(http://example.com/other-style.css);\\", \\"screen and print\\"]);
exports.push([module.id, \\"@import url(//example.com/style.css);\\", \\"\\"]);
exports.i(require(\\"-!../../../src/index.js??ref--4!package/test.css\\"), \\"\\");
exports.i(require(\\"-!../../../src/index.js??ref--4!./query.css?foo=1&bar=1\\"), \\"\\");
exports.i(require(\\"-!../../../src/index.js??ref--4!./other-query.css?foo=1&bar=1\\"), \\"\\");

// CSS module
exports.push([module.id, \\"@import url();\\\\n@import url('');\\\\n@import url(\\\\\\"\\\\\\");\\\\n@import '';\\\\n@import \\\\\\"\\\\\\";\\\\n@import \\\\\\"   \\\\\\";\\\\n@import url();\\\\n@import url('');\\\\n@import url(\\\\\\"\\\\\\");\\\\n@import ;\\\\n@import foo-bar;\\\\n@import-normalize;\\\\n@import url('http://') :root {}\\\\n\\\\n.class {\\\\n  a: b c d;\\\\n}\\\\n\\\\n.foo {\\\\n  @import 'path.css';\\\\n}\\\\n\\", \\"\\"]);
"
`;

exports[`import true: warnings 1`] = `
Array [
  [ModuleWarning: Module Warning (from \`replaced original path\`):
Warning

(10:1) Unable to find uri in '@import '''],
  [ModuleWarning: Module Warning (from \`replaced original path\`):
Warning

(11:1) Unable to find uri in '@import ""'],
  [ModuleWarning: Module Warning (from \`replaced original path\`):
Warning

(12:1) Unable to find uri in '@import "   "'],
  [ModuleWarning: Module Warning (from \`replaced original path\`):
Warning

(13:1) Unable to find uri in '@import url()'],
  [ModuleWarning: Module Warning (from \`replaced original path\`):
Warning

(14:1) Unable to find uri in '@import url('')'],
  [ModuleWarning: Module Warning (from \`replaced original path\`):
Warning

(15:1) Unable to find uri in '@import url("")'],
  [ModuleWarning: Module Warning (from \`replaced original path\`):
Warning

(24:1) Unable to find uri in '@import '],
  [ModuleWarning: Module Warning (from \`replaced original path\`):
Warning

(25:1) Unable to find uri in '@import foo-bar'],
  [ModuleWarning: Module Warning (from \`replaced original path\`):
Warning

(27:1) It looks like you didn't end your @import statement correctly. Child nodes are attached to it.],
  [ModuleWarning: Module Warning (from \`replaced original path\`):
Warning

(6:1) Unable to find uri in '@import url()'],
  [ModuleWarning: Module Warning (from \`replaced original path\`):
Warning

(7:1) Unable to find uri in '@import url('')'],
  [ModuleWarning: Module Warning (from \`replaced original path\`):
Warning

(8:1) Unable to find uri in '@import url("")'],
]
`;
